packages/primitives/popover/src/popover-root.directive.ts

```diff
@@ -142,20 +158,104 @@ export class RdxPopoverRootDirective implements OnInit {
+                    console.log(
+                        this.uniqueId(),
+                        '[onStateChangeEffect] pre filter',
+                        state,
+                        this.getAnimationParamsSnapshot()
+                    );
+                if (!this.ifRunOnStateChangeCallbackSync(state)) {
+                    return;
+                }
+                isRdxPopoverDevMode() && console.log(this.uniqueId(), '[onStateChangeEffect]', state);
+                this.onStateChangeCallback(state);
+            });
+        });
+    }
+
+    /** @ignore */
+    private onCssAnimationStatusChangeChangeEffect() {
         effect(() => {
-            const isOpen = this.isOpen();
+            const cssAnimationStatus = this.cssAnimationStatus();
 
             untracked(() => {
-                if (isOpen) {
-                    this.show();
-                } else {
-                    this.hide();
+                isRdxPopoverDevMode() &&
+                    console.log(
+                        this.uniqueId(),
+                        '[onCssAnimationStatusChangeChangeEffect pre filter]',
+                        cssAnimationStatus,
+                        this.getAnimationParamsSnapshot()
+                    );
+
+                if (!this.ifRunOnStateChangeCallbackAsync(cssAnimationStatus)) {
+                    return;
                 }
+
+                isRdxPopoverDevMode() &&
+                    console.log(
+                        this.uniqueId(),
+                        '[onCssAnimationStatusChangeChangeEffect]',
+                        cssAnimationStatus,
+                        this.state()
+                    );
+                this.onStateChangeCallback(this.state());
             });
         });
     }
 
+    /** @ignore */
+    private ifRunOnStateChangeCallbackSync(state: RdxPopoverState) {
+        return (
+            !this.popoverContentAttributesDirective() ||
+            !this.cssAnimation() ||
+            !this.cssOnCloseAnimation() ||
+            state === RdxPopoverState.OPEN
+        );
+    }
+
+    /** @ignore */
+    private ifRunOnStateChangeCallbackAsync(cssAnimationStatus: RdxPopoverAnimationStatus | null) {
+        return (
+            this.popoverContentAttributesDirective() &&
+            this.cssAnimation() &&
+            this.cssOnCloseAnimation() &&
+            cssAnimationStatus &&
+            this.state() === RdxPopoverState.CLOSED &&
+            [RdxPopoverAnimationStatus.CLOSED_ENDED].includes(cssAnimationStatus)
+        );
+    }
+
+    /** @ignore */
+    private onStateChangeCallback(state: RdxPopoverState) {
+        const isOpen = this.isOpen(state);
+        isOpen ? this.show() : this.hide();
+        this.onOpenChange.emit(isOpen);
+        this.document.dispatchEvent(
+            new CustomEvent(`popover.${isOpen ? 'open' : 'close'}`, {
+                detail: {
+                    id: this.name()
+                }
+            })
+        );
```

Comment:
@radix-ng/core @radix-ng/maintainers is this `this.document.dispatchEvent` a standard for this lib? I've copied it from the `tooltip` but I cannot see it being used anywhere else

Reply:
I think this is copy-pasted from the original React implementation. For Angular, it doesn’t seem necessary.

```tsx
const [open = false, setOpen] = useControllableState({
    prop: openProp,
    defaultProp: defaultOpen,
    onChange: (open) => {
      if (open) {
        providerContext.onOpen();

        // as `onChange` is called within a lifecycle method we
        // avoid dispatching via `dispatchDiscreteCustomEvent`.
        document.dispatchEvent(new CustomEvent(TOOLTIP_OPEN));
      } else {
        providerContext.onClose();
      }
      onOpenChange?.(open);
    },
  });
```

And dont forget about SSR support :)

Reply:
good point, will improve SSR in another PR :)